sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1404,6 +1404,15 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     }
   }
 
+  override def visitIdentifier(ctx: IdentifierContext): String = withOrigin(ctx) {
+    val keyword = ctx.getText
+    if (ctx.ansiReserved() != null) {
+      throw new ParseException(
+        s"'$keyword' is reserved and you cannot use this keyword as an identifier.", ctx)
```

Comment:
maybe a simpler change is, in the antlr file, remove `(ansi)? ansiReservedKeywords` from `identifiers`.

Reply:
AFAIK reserved keywords can't be used as identifiers for all kinds, like table name/alias, column name/alias, etc. Shall we follow?

Reply:
If we just remove the `(ansi)? ansiReservedKeywords` from `identifier`, all the rule using `identifier` can't accept reserved keywords. Is this expected?
For example, we already `current_timestamp`/`current_date` as bulit-in functions in `FunctionRegistry`, but these keywords are reserved when ansi=true.
https://github.com/apache/spark/blob/2d2fb34b93a55af7e45191c8974bb55a0ed76e87/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4#L529

Reply:
But this is the ansi way to get the current timestamp, isn't it?

`SELECT CURRENT_TIMESTAMP` is valid.
`SELECT CURRENT_TIMESTAMP()` is invalid.

Reply:
Oh, I see. I'll check if we can just remove the rule.

Reply:
Oh, I see. But, if we just remove the rule, the two cases throw a parse error since both colunm name rule and function name rule use `identifier`;
```
scala> sql("SET spark.sql.parser.ansi.enabled=false")
scala> sql("SELECT CURRENT_TIMESTAMP").show
+--------------------+                                                          
| current_timestamp()|
+--------------------+
|2019-02-25 16:26:...|
+--------------------+

scala> sql("SELECT CURRENT_TIMESTAMP()").show
+--------------------+                                                          
| current_timestamp()|
+--------------------+
|2019-02-25 16:26:...|
+--------------------+

scala> sql("SET spark.sql.parser.ansi.enabled=true")
scala> sql("SELECT CURRENT_TIMESTAMP").show
org.apache.spark.sql.catalyst.parser.ParseException:
no viable alternative at input 'CURRENT_TIMESTAMP'(line 1, pos 7)

== SQL ==
SELECT CURRENT_TIMESTAMP
-------^^^

scala> sql("SELECT CURRENT_TIMESTAMP()").show
rg.apache.spark.sql.catalyst.parser.ParseException:
no viable alternative at input 'CURRENT_TIMESTAMP'(line 1, pos 7)

== SQL ==
SELECT CURRENT_TIMESTAMP()
-------^^^
```
To be honest, I'm not 100% sure about what the ANSI reserved means.
For example, both postgresql and mysql reserve `current_timestamp`, but the behviour is different;
```
postgres=# SELECT CURRENT_TIMESTAMP;
              now              
-------------------------------
 2019-02-25 16:44:26.320698+09
(1 row)

postgres=# SELECT CURRENT_TIMESTAMP();
ERROR:  syntax error at or near ")"
LINE 1: SELECT CURRENT_TIMESTAMP();
                                 ^
mysql> SELECT CURRENT_TIMESTAMP;
+---------------------+
| CURRENT_TIMESTAMP   |
+---------------------+
| 2019-02-25 16:45:02 |
+---------------------+

mysql> SELECT CURRENT_TIMESTAMP();
+---------------------+
| CURRENT_TIMESTAMP() |
+---------------------+
| 2019-02-25 16:45:04 |
+---------------------+
```

Reply:
ah good point!

But if don't forbid CURRENT_TIMESTAMP as column name, `SELECT CURRENT_TIMESTAMP FROM t` is ambiguous.

It looks to me that we should create an entry for `CURRENT_TIMESTAMP` directly in the antlr file, and return the corresponding function for it in `AstBuilder.scala`

Reply:
ok, I'll try to fix in that way. Thanks!

Reply:
@cloud-fan How about the latest fix?